src/util/schnorr.rs

```diff
@@ -17,3 +17,64 @@
 //!
 
 pub use secp256k1::schnorrsig::{PublicKey, KeyPair};
+use secp256k1::{Secp256k1, Verification};
+use hashes::{Hash, HashEngine};
+use util::taproot::{TapBranchHash, TapTweakHash};
+
+/// Untweaked Schnorr public key
+pub struct UntweakedPublicKey(pub PublicKey);
+
+/// Tweaked Schnorr public key
+pub struct TweakedPublicKey(PublicKey);
+
+impl UntweakedPublicKey {
+    /// Tweaks an untweaked public key given an untweaked key and optional script tree merkle root.
+    /// 
+    /// This is done by using the equation Q = P + H(P|c)G, where 
+    ///  * Q is the tweaked key
+    ///  * P is the internal key
+    ///  * H is the hash function
+    ///  * c is the commitment data
+    ///  * G is the generator point
+    pub fn tweak<C: Verification>(&self, secp: Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> TweakedPublicKey {
+        //PR #677 has a function `TapTweakHash::from_key_and_tweak(..)` which can be simpler to use here.
+        
+        //Commitment data.
+        let h: Vec<u8> = match merkle_root {
+            Some(hash) => hash.to_vec(),
+            None => vec![]
+        };
+
+        //Extend the internal key by the commitment data and HashTapTweak it.
+        let mut tweak: Vec<u8> = self.0.serialize().to_vec();
+        tweak.extend_from_slice(&h);
+        let mut engine = TapTweakHash::engine();
+        engine.input(&tweak);
+        let tweak_value: [u8; 32] = TapTweakHash::from_engine(engine).into_inner();
+            
+
+        //Tweak the internal key by the hashtaptweak
+        let mut taptweaked_key = self.0.clone();
+        let parity = taptweaked_key.tweak_add_assign(&secp, &tweak_value).expect("Tweak failed");
+        if self.0.tweak_add_check(&secp, &taptweaked_key, parity, tweak_value) {
+            return TweakedPublicKey(taptweaked_key);
+        } else { panic!("Tweak Failed") }
```

Comment:
I think this is not needed

Reply:
How we can have an incorrect result if it is checked in line 58 already and this check adds nothing to that?

Reply:
https://doc.rust-lang.org/std/macro.unreachable.html ?